truffleruby/src/main/java/org/truffleruby/interop/java/JavaUtilitiesNodes.java

```diff
@@ -0,0 +1,179 @@
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.TruffleOptions;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.CreateCast;
+import com.oracle.truffle.api.dsl.NodeChild;
+import com.oracle.truffle.api.dsl.NodeChildren;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.Node.Child;
+import com.oracle.truffle.api.object.DynamicObject;
+import com.oracle.truffle.api.profiles.BranchProfile;
+
+@CoreClass("Truffle::Interop::Java")
+public class JavaUtilitiesNodes {
+
+    @CoreMethod(names = "to_ruby_string", required = 1, isModuleFunction = true)
+    @NodeChild(value = "args", type = RubyNode[].class)
+    public static abstract class ToRubyStringNode extends CoreMethodNode {
+	@Specialization
+	public DynamicObject toRubyString(String target) {
+	    return createString(getRope(target));
+	}
+
+	protected Rope getRope(String value) {
+	    return StringOperations.encodeRope(value, UTF8Encoding.INSTANCE);
+	}
+    }
+
+    @CoreMethod(names = "java_class_by_name", isModuleFunction = true, required = 1)
+    @NodeChild(value = "name", type = RubyNode.class)
+    public abstract static class JavaClassByNameNode extends CoreMethodNode {
+        @CreateCast("name")
+        public RubyNode coercetNameToString(RubyNode newName) {
+            return NameToJavaStringNodeGen.create(newName);
+        }
+
+        @Specialization
+        public Object javaClassByName(String name,
+                                      @Cached("create()") BranchProfile errorProfile) {
+            if ( ! TruffleOptions.AOT ) {
+                try {
+                    Class<?> klass =  Class.forName(name);
+                    return klass;
+                } catch (Exception e) {
+                    errorProfile.enter();
+                    throw new RaiseException(coreExceptions().nameErrorImportNotFound(name, this));
+                }
+            } else {
+                throw new RaiseException(coreExceptions().runtimeError("Not aavailable on SVM.", this));
+            }
+        }
+    }
+
+    @CoreMethod(names = "to_java_string", required = 1, isModuleFunction = true)
+    @NodeChildren({
+            @NodeChild(value = "target", type = RubyNode.class)
+    })
+    public static abstract class ToJavaStringNode extends CoreMethodNode {
+        @Specialization
+        public String toJavaString(VirtualFrame frame, DynamicObject string,
+                                   @Cached("create()") NameToJavaStringNode toJavaStringNode) {
+            return toJavaStringNode.executeToJavaString(frame, string);
+        }
+    }
+
+    @CoreMethod(names = "java_hash", required = 1, isModuleFunction = true)
+    @NodeChildren({
+            @NodeChild(value = "target", type = RubyNode.class)
+    })
+    public static abstract class JavaHashNode extends CoreMethodNode {
+        @Specialization
```

Comment:
No because `java_eql?` is just doing a reference comparison which is OK to compile (you would need it if it called `equals`).
The problem with `hashCode()` like this is you would potentially compile all possible implementations of `hashCode()` during partial evaluation, which is not useful.